refactor(share/availability): simplify light availability

walldiss opened the pull request:
- simplification refactoring of Samples selection code
- remove squareSampler
- remove binary encode/decode in favour of json
- removal of no longer relevant TODOs
- rename "cache" naming artefacts from the past
- use height in logs instead of root hash

Wondertan commented on share/availability/light/availability.go:
```diff
@@ -130,7 +126,10 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
 
 	// store the result of the sampling session
-	bs := encodeSamples(failedSamples)
+	bs, err := json.Marshal(failedSamples)
```
Why do we break the serialization? We save a few loc for a much slower and bloated serialization. I don't think that's worth it at all.

walldiss replied:
json encoding speed is 100-1000mb/s. It would take for our ~100 bytes encoding around 1ns. Does not seems to be a performance hit to me.

Wondertan replied:
The point is, there is no good reason to do it in the first place. Also, we don't need the JSON bloat.

walldiss replied:
As you pointed out in original comment, the benefit is simplified code. Custom encoding does not bring much of performance benefit here.

Wondertan replied:
It does being storage overhead tho

Wondertan replied:
Discussed in sync that breaking is fine:
* We don't store successful samples, so the only way to prune those is by resyncing, which is equivalent to breaking. 
* We are still allowed to break

Wondertan replied:
Another missed thing is that the store state is changing in https://github.com/celestiaorg/celestia-node/pull/3887 to contain successful samples. We could store two states separately to not break, but again, it still requires resyncing to cleanup old samples.

Wondertan commented on share/availability/light/sample.go:
```diff
@@ -1,104 +1,41 @@
 
-// Sample is a point in 2D space over square.
+// Sample represents a coordinate in a 2D data square.
 type Sample struct {
```
Let's make it private while we are on it. Can't recall it being used elsewhere. Please check

walldiss replied:
It would be used externally by `pruner` pkg very soon.

Wondertan replied:
Kk

renaynay reviewed (commented):
Additional request is to move dah validation check in `availability.go` earlier in the `SharesAvailability` function -- ideally in the beginning. It doesn't make sense where it currently is.

walldiss commented:
Had to rebase the PR on main, and commit history is lost. Here is what changed after last reviews
- Removed uneccesary dah validation from both light and full availability
- clean up remaining logs to use height.
- changed uint -> int for Row and Col in Sample
